src/Psalm/Internal/Analyzer/Statements/Expression/BinaryOp/ConcatAnalyzer.php

```diff
@@ -277,6 +278,41 @@ public static function analyze(
                     }
                 }
             }
+        } elseif ($left_type || $right_type) {
+            /**
+             * @var Union $known_operand
+             */
+            $known_operand = $right_type ?? $left_type;
+
+            $numeric_type = new Union([
+                new TNumericString,
+                new TInt,
+                new TFloat,
+            ]);
+
+            $non_empty_string = $numeric_type->getBuilder()->addType(new TNonEmptyString())->freeze();
+
+            $non_falsy_string = $numeric_type->getBuilder()->addType(new TNonFalsyString())->freeze();
+
+            $known_operand_not_empty = UnionTypeComparator::isContainedBy(
```

Comment:
You can use something like 
`if($known_operand->isSingle() && $known_operand->getSingleAtomic() instanceof TNonEmptyString)` to check for the type in the Union, it will be quicker